kernel/src/transaction/mod.rs

```diff
@@ -295,32 +315,58 @@ impl Transaction {
             ));
         }
 
-        // validate domain metadata
-        let mut domains = HashSet::new();
-        for domain_metadata in &self.domain_metadatas {
-            if domain_metadata.is_internal() {
+        // validate user domain metadata and check if we have removals
+        let mut seen_domains = HashSet::new();
+        let mut has_removals = false;
+        for dm in &self.domain_metadatas {
+            if dm.is_internal() {
                 return Err(Error::Generic(
                     "Cannot modify domains that start with 'delta.' as those are system controlled"
                         .to_string(),
                 ));
             }
-            if !domains.insert(domain_metadata.domain()) {
+
+            if !seen_domains.insert(dm.domain()) {
                 return Err(Error::Generic(format!(
                     "Metadata for domain {} already specified in this transaction",
-                    domain_metadata.domain()
+                    dm.domain()
                 )));
             }
+
+            if dm.is_removed() {
+                has_removals = true;
+            }
         }
 
+        // fetch previous configuration values (requires log replay)
+        let existing_domains = if has_removals {
+            scan_domain_metadatas(self.read_snapshot.log_segment(), None, engine)?
+        } else {
+            HashMap::new()
+        };
+
+        let user_domains = self
+            .domain_metadatas
+            .iter()
+            .filter_map(move |dm: &DomainMetadata| {
+                if dm.is_removed() {
+                    existing_domains.get(dm.domain()).map(|existing| {
+                        DomainMetadata::remove(
+                            dm.domain().to_string(),
+                            existing.configuration().to_string(),
+                        )
+                    })
+                } else {
+                    Some(dm.clone())
+                }
+            });
```

Comment:
We can consume the domain_metadatas and avoid the clone since it seems like we don't use it anymore
```suggestion
        let user_domains = self
            .domain_metadatas
            .into_iter()
            .filter_map(move |dm: DomainMetadata| {
                if !dm.is_removed() {
                    return Some(dm);
                }
                existing_domains.get(dm.domain()).map(|existing| {
                    DomainMetadata::remove(
                        dm.domain().to_string(),
                        existing.configuration().to_string(),
                    )
                })
            });
```

Reply:
If we decide to use `removed_domains` this becomes:
We can consume the domain_metadatas and avoid the clone since it seems like we don't use it anymore
```suggestion
        let user_domains = self
            .removed_domains
            .iter()
            .map(move |domain_name: &str| {
                existing_domains.get(domain_name).map(|existing| {
                    DomainMetadata::remove(
                        dm.domain().to_string(),
                        existing.configuration().to_string(),
                    )
                })
            })
            .chain(domain_metadatas);
```

Reply:
will probably apply suggestion 1 or 2 depending on our takeaways from the other threads :)

Reply:
oof I'm also realizing that because "generate_domain_metadata" can be called multiple times, the `into_iter` doesn't work. Also given the deferred error handling, I think removes can remain as they are for now.